### PR TITLE
New changes

### PR DESCRIPTION
**Pull request by ZgzInfinity**
Control de number of visits of each user and implemented a social bar to share the website

### CODE REVIEW

File: src/main/java/es/unizar/webeng/hello/HelloController.java [hunk 1]
```diff
@@ -78,6 +80,25 @@ public String welcome(Map<String, Object> model) {
             addr = InetAddress.getLocalHost();
             /** Sets "hostname" attribute to server machine name */
             model.put("hostname", addr.getHostName());
+
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/HelloController.java]: This method starts to grow. Refactor your code and move the control to a method.

File: src/main/java/es/unizar/webeng/hello/RegisterIP.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	
+	
+	public RegisterIP() {
+	}
+
+	public InetAddress getDirection() {
+		return direction;
+	}
+
+	public void setDiretion(InetAddress direction) {
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/RegisterIP.java]: Fix the typo.

File: src/main/java/es/unizar/webeng/hello/RegisterIP.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	public int getNumberVisits() {
+		return numberVisits;
+	}
+
+	public void setNumberVisits(int numberVisits) {
+		this.numberVisits = numberVisits;
+	}
+
+	@Override
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/RegisterIP.java]: Use the utility class Objects. See example in [here](https://kodejava.org/how-do-i-implement-equals-and-hashcode-method-using-java-util-objects/)

File: src/main/java/es/unizar/webeng/hello/HelloController.java [hunk 4]
```diff
@@ -78,6 +80,25 @@ public String welcome(Map<String, Object> model) {
             addr = InetAddress.getLocalHost();
             /** Sets "hostname" attribute to server machine name */
             model.put("hostname", addr.getHostName());
+            
+            /** Control of number of visits for each user **/
+            
+            RegisterIP r = new RegisterIP();
+            r.setDiretion(addr);
+            
+            int index = containsRegisterIP(r);
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/HelloController.java]: Concurrent access is possible from the same IP. Worst case, two threads with index = -1 for the same IP

File: src/main/java/es/unizar/webeng/hello/HelloController.java [hunk 5]
```diff
@@ -78,6 +80,25 @@ public String welcome(Map<String, Object> model) {
...
+           
+            if (index == -1) {
+            	r.setNumberVisits(1);
+            	listaIps.addLast(r);
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/HelloController.java]: Due to the previous case, the same IP may be inserted twice.

File: src/main/java/es/unizar/webeng/hello/HelloController.java [hunk 6]
```diff
@@ -78,6 +80,25 @@ public String welcome(Map<String, Object> model) {
...
+            }
+            else {
+            	r = listaIps.get(index);
+            	int times = r.getNumberVisits() + 1;
```
Comment (fjlopez) [inline src/main/java/es/unizar/webeng/hello/HelloController.java]: Potential race case, the same IP can be handled by two threads, losing one of the visits.

File: src/main/java/es/unizar/webeng/hello/RegisterIP.java (hunk 3 above)
Comment (ZgzInfinity) [reply src/main/java/es/unizar/webeng/hello/RegisterIP.java]: Does this comment mean that i only have to change the equals and hashcode methods?

File: src/main/java/es/unizar/webeng/hello/HelloController.java (hunk 1 above)
Comment (ZgzInfinity) [reply src/main/java/es/unizar/webeng/hello/HelloController.java]: ok i move the code to another method which is called monitorizeNumberOfVisits(addr) where addr is the IP of the user. So i'll invoke it in the method wellcome.

File: src/main/java/es/unizar/webeng/hello/HelloController.java (hunk 4 above)
Comment (ZgzInfinity) [reply src/main/java/es/unizar/webeng/hello/HelloController.java]: ok so if i reference the method with the keyword synchronized this problem should be solve if
i'm not wrong.

File: src/main/java/es/unizar/webeng/hello/RegisterIP.java (hunk 3 above)
Comment (fjlopez) [reply src/main/java/es/unizar/webeng/hello/RegisterIP.java]: Yes, it is simpler and less error prone.

File: src/main/java/es/unizar/webeng/hello/HelloController.java (hunk 4 above)
Comment (fjlopez) [reply src/main/java/es/unizar/webeng/hello/HelloController.java]: Yep. Will depend on how you implement it, but this it seems ok.

**Comment by fjlopez** [comment]
I have a doubt. Why does your forked repository appear as unknown?
image

**Pull request closed without merging**